### PR TITLE
fix mul/div overflow issue on CPU float16

### PR DESCRIPTION
**Pull request by mingfeima**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #98820

Fix https://github.com/pytorch/pytorch/issues/63482 and https://github.com/pytorch/pytorch/issues/98691

The above two issues have the same root cause:

**binary_ops** will create TensorIterator with the flag `promote_inputs_to_common_dtype` on, which will convert both input tensors to the common_dtype_ (the logic is bypassed on CUDA), which might overflow on Half. If one of the inputs is a scalar with abs value larger than ~65000, it will overflow.

This patch will try to fetch the scalar value from the `original_tensor_base` which records the original scalar input value, then in the `cpu_kernel_vec` the TensorIterator is treated as an unary Op.

So previously, CPU and CUDA would have different behaviors for such scenario. This is aligned with this patch, test cases added for both CPU and CUDA device.

The following is the results:

#### before:
```
>>> torch.tensor([3388.], dtype=torch.half).div(524288.0)
tensor([0.], dtype=torch.float16)
>>> torch.tensor([0.01], dtype=torch.float16) * torch.tensor(65536, dtype=torch.float32)
tensor([inf], dtype=torch.float16)
```

#### after:
```
>>> torch.tensor([3388.], dtype=torch.half).div(524288.0)
tensor([0.0065], dtype=torch.float16)
>>> torch.tensor([0.01], dtype=torch.float16) * torch.tensor(65536, dtype=torch.float32)
tensor([655.5000], dtype=torch.float16)
```

Also need to update `RRelu` implementation, to use float to store the intermediate results, otherwise the following test case would fail:
```
. build/bin/test_api --gtest_filter=ModulesTest.RReLU
```

cc @jgong5 @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

File: aten/src/ATen/native/cpu/BinaryOpsKernel.cpp [hunk 1]
```diff
@@ -147,6 +191,57 @@ void div_floor_kernel(TensorIteratorBase& iter) {
...
     });
   } else {
+    // See NOTE: [Floor Division in Python]
+    if (iter.is_scalar(2) && at::isReducedFloatingType(dtype)) {
+      AT_DISPATCH_REDUCED_FLOATING_TYPES(dtype, "div_floor_cpu_reduced_float", [&]() {
+        float b = iter.original_scalar_value<float>(2);
+        iter.remove_operand(2);
+        using vec_t = Vectorized<float>;
+        cpu_kernel_vec(iter,
+          [=](scalar_t aa) __ubsan_ignore_float_divide_by_zero__ -> scalar_t {
```
Comment (jgong5) [inline aten/src/ATen/native/cpu/BinaryOpsKernel.cpp:201]: Can we share the logic instead of copying code from https://github.com/pytorch/pytorch/pull/98820/files#diff-0fd0c99904c62a1a7e46b08568d047c4e9ab153b529a8c6a2317ac6a3726bdffR249?

File: aten/src/ATen/native/cpu/BinaryOpsKernel.cpp [hunk 2]
```diff
@@ -147,6 +191,57 @@ void div_floor_kernel(TensorIteratorBase& iter) {
...
+              floordiv = std::floor(div);
+              if (div - floordiv > scalar_t(0.5)) {
+                floordiv += scalar_t(1.0);
+              }
+            } else {
+              floordiv = c10::copysign(scalar_t(0), a / b);
+            }
+            return floordiv;
+          },
+          [=](Vectorized<scalar_t> a) {
```
Comment (jgong5) [inline aten/src/ATen/native/cpu/BinaryOpsKernel.cpp:225]: ditto: share code?

File: aten/src/ATen/native/cpu/BinaryOpsKernel.cpp
Comment (jgong5) [inline aten/src/ATen/native/cpu/BinaryOpsKernel.cpp:1]: How about other kernels like `remainder_kernel`?

**Comment by ngimel** [comment]
>It might even make sense for torch.mul(float16 tensor, float32 tensor) to return a float32 tensor while torch.mul(float16 tensor, float32 scalar) to return a float16 tensor...

This is exactly what's happening today, the issue being fixed here is over/underflowing of intermediate results, whiche is separate from dtypes of inputs and outputs.

**Comment by vadimkantorov** [comment]
> This is exactly what's happening today,

Then I guess those type promotion examples would benefit from more explicit examples, e.g with all of:

torch.ones(1) - 1-dim, 1-numel - the only case in existing examples
torch.ones(2) - 1-dim, 2-numel
torch.tensor(x) - 0-dim, 1-numel

**Comment by mingfeima** [comment]
> 
> This is exactly what's happening today, the issue being fixed here is over/underflowing of intermediate results, whiche is separate from dtypes of inputs and outputs.

@vadimkantorov: as @ngimel commented, this patch is to fix the overflow with scalar inputs, yet I think a more clear doc would be very helpful.

From our angle, we try to align the CPU behavior on `Half` with CUDA device.
As you can see: previously, `torch.tensor([3388.], dtype=torch.half) / 524288` would give 0 on CPU and 0.0065 on CUDA. This type behavior would be very confusing for users, and give unexpected results (usually very difficult to debug).

File: aten/src/ATen/native/cpu/BinaryOpsKernel.cpp (hunk 1 above)
Comment (mingfeima) [reply aten/src/ATen/native/cpu/BinaryOpsKernel.cpp]: updated~

**Review by jgong5 — APPROVED** [review]

**Comment by mingfeima** [comment]
@ngimel Could you please help review this one ?

**Comment by vadimkantorov** [comment]
> So yeah, for docs it would be interesting to know the case: torch.mul(1-dim 1-numel tensor, 0-dim 1-numel tensor) -> does it behave like torch.mul(1-dim 2-numel, 0-dim 1-numel) or like torch.mul(0-dim 1-numel, 0-dim 1-numel)
> 
> If the behavior suddenly switches it can lead to "return type" instability depending on how many elements has the first 1-dim argument

at least it would be nice to add these cases as tests...

File: aten/src/ATen/native/Activation.cpp [hunk 3]
```diff
@@ -576,8 +577,9 @@ inline void _rrelu_with_noise_train(
     const Scalar& lower_,
     const Scalar& upper_,
     c10::optional<Generator> generator) {
-  scalar_t lower = lower_.to<scalar_t>();
-  scalar_t upper = upper_.to<scalar_t>();
+  using acc_t = at::opmath_type<scalar_t>;
```
Comment (ngimel) [inline aten/src/ATen/native/Activation.cpp:580]: nit: here (and in all your half PRs) please use `opmath_t`, not `acc_t`

File: aten/src/ATen/native/cpu/BinaryOpsKernel.cpp [hunk 4]
```diff
@@ -55,18 +61,29 @@ void atan2_kernel(TensorIteratorBase& iter) {
...
     cpu_kernel(
         iter,
         [=](c10::complex<at::Half> a,
             c10::complex<at::Half> b) -> c10::complex<at::Half> {
           using comp_t = c10::complex<float>;
           return comp_t{a} * comp_t{b};
         });
+  } else if (iter.is_scalar(2) && at::isReducedFloatingType(dtype)) {
+    AT_DISPATCH_REDUCED_FLOATING_TYPES(dtype, "mul_cpu_reduced_float", [&]() {
+      float b = iter.original_scalar_value<float>(2);
```
Comment (ngimel) [inline aten/src/ATen/native/cpu/BinaryOpsKernel.cpp:77]: nit: don't hardcode `float`, use `opmath_t`

File: aten/src/ATen/native/cpu/BinaryOpsKernel.cpp [hunk 5]
```diff
@@ -55,18 +61,29 @@ void atan2_kernel(TensorIteratorBase& iter) {
...
-  } else if (iter.dtype() == kComplexHalf) {
+  } else if (dtype == kComplexHalf) {
     cpu_kernel(
         iter,
         [=](c10::complex<at::Half> a,
             c10::complex<at::Half> b) -> c10::complex<at::Half> {
           using comp_t = c10::complex<float>;
           return comp_t{a} * comp_t{b};
         });
+  } else if (iter.is_scalar(2) && at::isReducedFloatingType(dtype)) {
```
Comment (ngimel) [inline aten/src/ATen/native/cpu/BinaryOpsKernel.cpp:75]: why is it only `is_scalar(2)`, is `mul` constructed in such a way that scalar argument is the second?

File: test/test_binary_ufuncs.py [hunk 6]
```diff
@@ -3026,6 +3026,14 @@ def test_muldiv_scalar(self, device, dtype):
         self.assertEqual(x / s, x / y)
         self.assertEqual(s / x, y / x)
 
+        if dtype == torch.half:
```
Comment (ngimel) [inline test/test_binary_ufuncs.py:3029]: there's `test_divmul_scalar` that runs only on cuda currently that you should be able to make run on cpu, and not change this test?

File: aten/src/ATen/native/Activation.cpp (hunk 3 above)
Comment (mingfeima) [reply aten/src/ATen/native/Activation.cpp]: updated!

File: aten/src/ATen/native/cpu/BinaryOpsKernel.cpp (hunk 4 above)
Comment (mingfeima) [reply aten/src/ATen/native/cpu/BinaryOpsKernel.cpp]: updated!

File: aten/src/ATen/native/cpu/BinaryOpsKernel.cpp (hunk 5 above)
Comment (mingfeima) [reply aten/src/ATen/native/cpu/BinaryOpsKernel.cpp:77]: TensorIterator has logic to purmute scalar input to the rear, so this patch will work for both of the following senarios:
```
>>> torch.tensor([0.01], dtype=torch.float16) * 1e5
tensor([1000.], dtype=torch.float16)
>>> 1e5 * torch.tensor([0.01], dtype=torch.float16)
tensor([1000.], dtype=torch.float16)
```

File: test/test_binary_ufuncs.py (hunk 6 above)
Comment (mingfeima) [reply test/test_binary_ufuncs.py]: Oh, thanks for the remainder, did not notice this one before.

**Review by ngimel — APPROVED** [review]
